tests/test_salcommander_minimal_controllers.py

```diff
@@ -0,0 +1,145 @@
+
+import subprocess
+import pathlib
+import unittest
+import warnings
+
+from lsst.ts import salobj
+
+# Long enough to perform any reasonable operation
+# including starting a CSC or loading a script (seconds)
+STD_TIMEOUT = 60
+
+INITIAL_LOG_LEVEL = 20
+
+index_gen = salobj.index_generator()
+
+languages = ["cpp", "java"]
+
+
+class SalCommanderToMinimalControllerTestCase(unittest.TestCase):
+    """Test a C++ or Java Commander communicating to
+    various minimal controllers."""
+
+    def setUp(self) -> None:
+        salobj.set_random_lsst_dds_partition_prefix()
+        self.index = next(index_gen)
+
+    def getLanguageVariables(self, language) -> None:
+        """Get the language specific test values;
+        either cpp (C++) or java (Java).
+
+        Parameters
+        ----------
+        language : `str`
+            Options are "java" or "cpp" and are case insensitive.
+
+        """
+        if language.lower() == "cpp":
+            commander = "minimal_cpp_commander.sh"
+            language_upper = "C++"
+            initial_log = "initial_log_level="
+        elif language.lower() == "java":
+            commander = "minimal_java_commander.sh"
+            language_upper = "Java"
+            initial_log = "initial logLevel.level "
+        else:
+            raise ValueError(f"Argument {language} must be one of 'cpp' or 'java'")
+        return commander, language_upper, initial_log
+
+    def test_cpp_controller(self) -> None:
+        for language in languages:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_cpp_controller.sh")
+
+    def test_java_controller(self) -> None:
+        for language in languages:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_java_controller.sh")
+
+    def test_salobj_controller(self) -> None:
+        for language in languages:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_salobj_controller.py")
+
+    def check_minimal_controller(self, language: str, exec_name: str) -> None:
+        # Start the Commander before the Controller subprocess
+        # to be sure the Commander sees telemetry from the Controller
+        # (telemetry is volatile, so has no historical data).
+        commander, language_upper, initial_log = self.getLanguageVariables(language)
+        commander_path = pathlib.Path(__file__).home() / "repos/ts_sal/bin" / commander
+        print(f"{language_upper} Commander: start {commander_path} with index={self.index} in a subprocess")
+        assert commander_path.is_file()
+        commander_process = subprocess.Popen(
+            [
+                commander_path,
+                str(self.index),
+                str(INITIAL_LOG_LEVEL),
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+        )
```

Comment:
You are careful to make sure controller_process is cleaned up, which is great. Please do the same for commander_process.

It turns out that subprocess.Popen can be used as a "context manager" in a "with" statement, which does cleanup automatically. There's no need for try/finally. All you have to do is:
```
        with subprocess.Popen(
            [commander_path, str(self.index), str(INITIAL_LOG_LEVEL)],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
        ) as commander_process, subprocess.Popen(
            [str(controller_path), str(self.index), str(INITIAL_LOG_LEVEL)],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
        ) as controller_process:
            print(f"{language_upper} Commander: wait for completion")
            commander_process.wait()
            ...
```

Reply:
Fixed